### DISCUSSION TITLE
Sensitivity map Error when running adjoint [Rotor 37]

### DISCUSSION THREAD

**Discussion by Mohsin-Bashir** [mdolab/dafoam#664] [General]
I have added `"writeSensMap": ["shapey", "shapez"],` to the runScript_v2.py file of rotor 37 tutorial but when I run it I am getting error AssertionError: invalid array size!. I have attached my log file  [logAdjoint.txt](https://github.com/user-attachments/files/16358515/logAdjoint.txt)

![image](https://github.com/user-attachments/assets/473cee15-d9bc-4ed0-88d2-43b17537869f)

**Comment by friedenhe** [comment]
Can you set it to `"writeSensMap": ["shapey"]`?

**Reply by Mohsin-Bashir** [reply]
tried but same issue

**Comment by friedenhe** [comment]
This could be a potential bug for the sensitivity map. We require that the design surfaces and the all-wall surfaces are the same. Can you confirm that your design surface "blade" is the only wall surface in your case? Do you have other wall surfaces, such as hub?

**Reply by Mohsin-Bashir** [reply]
> This could be a potential bug for the sensitivity map. We require that the design surfaces and the all-wall surfaces are the same. Can you confirm that your design surface "blade" is the only wall surface in your case? Do you have other wall surfaces, such as hub?

Would it be possible to modify the DAFoam code so that the design surfaces and the all-wall surfaces don't have to be the same? This way, I can simply add one design surface, namely the blade, and run the optimization with a sensitivity map focused solely on the blade. If possible, it would simplify many things for me.

**Reply by friedenhe** [reply]
I believe the design and all-wall surfaces don't have to be the same in the current version of DAFoam...

**Reply by Mohsin-Bashir** [reply]
Thank you so much. Actually I hadn't noticed the changes.

**Comment by Mohsin-Bashir** [comment]
Blade, Hub and shroud are all of type wall. How should I change the boundary file to calculate sensitivity map? (btw this is the tutorial case for rotor37)
```
7
(
    blade
    {
        type            wall;
        inGroups        1(wall);
        nFaces          2670;
        startFace       121224;
    }
    shroud
    {
        type            wall;
        inGroups        1(wall);
        nFaces          2826;
        startFace       123894;
    }
    hub
    {
        type            wall;
        inGroups        1(wall);
        nFaces          2826;
        startFace       126720;
    }
    inlet
    {
        type            patch;
        nFaces          750;
        startFace       129546;
```

**Comment by Mohsin-Bashir** [comment]
I tried changing the type of hub and shroud to "patch" but then I'm getting the following error 
```
Selecting turbulence model type RAS
Selecting RAS turbulence model SpalartAllmarasFv3
[3]
[3]
[3] --> FOAM FATAL ERROR:
[3] Invalid wall function specification
    Patch type for patch shroud must be wall
    Current patch type is genericPatch
```

**Comment by friedenhe** [comment]
I meant to add hub and shroud as your designSurfaces and create a new (larger) FFD to include blade, hub, and shroud.

**Reply by Mohsin-Bashir** [reply]
I have added hub and shroud as designSurfaces and also used a bigger FFD globalFFD.xyz as is given in tutorial [runScript_v2.txt](https://github.com/user-attachments/files/16558843/runScript_v2.txt), but am still getting the same error [logAdjoint.txt](https://github.com/user-attachments/files/16558842/logAdjoint.txt).

**Comment by friedenhe** [comment]
Can you change this line https://github.com/mdolab/dafoam/blob/21dc6b03ce30f41578733e19d3bbbde06d3b7634/dafoam/pyDAFoam.py#L2074 to `Xs = self.getSurfaceCoordinates(self.designSurfacesGroup)`, then run `pip install .`, then, run the case again? If this fix works for you, I will update DAFoam.

**Reply by Mohsin-Bashir** [reply]
Thank you, it's working now. 
Please update the Rotor 37 tutorial to include the following points: 
ensure that the hub and shroud are added as design surfaces, and use a larger FFD
